File: platform/src/main/java/ru/abondin/hreasy/platform/repo/vacation/VacationRepo.java
package ru.abondin.hreasy.platform.repo.vacation;

import org.springframework.data.r2dbc.repository.Query;
import org.springframework.data.repository.reactive.ReactiveCrudRepository;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Flux;
import ru.abondin.hreasy.platform.repo.manager.ManagerRepo;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.List;

@Repository
public interface VacationRepo extends ReactiveCrudRepository<VacationEntry, Integer> {

    @Query("select e.display_name as employee_display_name" +
            ", e.email as employee_email" +
            ", e.current_project as employee_current_project" +
            ", e.current_project_role as employee_current_project_role" +
            ", p.name as employee_current_project_name" +
            ", v.* from  vac.vacation v " +
            " inner join empl.employee e on e.id=v.employee" +
            " left join proj.project p on e.current_project=p.id" +
            " where v.year in (:years)" +
            " and e.date_of_dismissal is null or e.date_of_dismissal > :now" +
            " order by v.end_date asc")
    Flux<VacationView> findAll(List<Integer> years, OffsetDateTime now);

    /**
     * @param employeeId
     * @param now
     * @return
     */
    @Query("select v.* from vac.vacation v " +
            "where v.employee=:employeeId and (v.start_date>=:now or v.end_date>=:now)" +
            " order by v.end_date asc")
    Flux<VacationEntry> findFuture(Integer employeeId, OffsetDateTime now);


    @Query("""
            select e.display_name as employee_display_name
            , e.email as employee_email
            , e.current_project as employee_current_project
            , p.name as employee_current_project_name
            , m.emails as managers_emails
            , v.* from  vac.vacation v 
             inner join empl.employee e on e.id=v.employee
             left join proj.project p on e.current_project=p.id
             left join
             (
             """ + ManagerRepo.EMPLOYEE_ALL_MANAGER_EMAILS +
            """
                     ) m on m.id=e.id
                     where v.year = :year
                     and (e.date_of_dismissal is null or e.date_of_dismissal > :now)
                     and v.start_date>=:from and v.start_date<=:to and v.stat in (0)
                     order by v.end_date asc
                    """)
    Flux<VacationViewWithManagers> findPlannedStartedBetween(OffsetDateTime now, int year, LocalDate from, LocalDate to);
}

File: platform/src/main/kotlin/ru/abondin/hreasy/platform/repo/VacationRepo.kt
package ru.abondin.hreasy.platform.repo

import org.springframework.data.annotation.Id
import org.springframework.data.r2dbc.core.DatabaseClient
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import java.time.LocalDate


@Table("vacation")
data class VacationEntry(
        @Id
        val id: Int,
        val employee: Int,
        var year: Int,
        @Column("start_date")
        var startDate: LocalDate?,
        @Column("end_date")
        var endDate: LocalDate?,
        var notes: String?
)

data class VacationView(
        val id: Int,
        val employee: Int,
        @Column("employee_firstname")
        val employeeFirstname: String?,
        @Column("employee_lastname")
        val employeeLastname: String?,
        @Column("employee_patronymic_name")
        val employeePatronymicName: String?,
        var year: Int,
        var start_date: LocalDate?,
        var end_date: LocalDate?,
        var notes: String?
)

interface VacationRepo {
    fun findAll(startsFrom: LocalDate): Flux<VacationView>;
}

@Service
class VacationRepoImpl(private val databaseClient: DatabaseClient) : VacationRepo {

    override fun findAll(startsFrom: LocalDate): Flux<VacationView> {
        var sql = "select v.*, e.firstname as employee_firstname, e.lastname as employee_lastname, e.patronymic_name as employee_patronymic_name" +
                " from vacation v inner join employee e on e.id=v.employee" +
                " where v.start_date>=:startsFrom"
        " order by v.start_date asc";
        val select = databaseClient
                .execute(sql)
                .bind("startsFrom", startsFrom)
                .`as`(VacationView::class.java);
        return select.fetch().all();
    }
}